src/web/security/core/src/main/java/org/geoserver/security/web/SecurityHomePageContentProvider.java

```diff
@@ -46,6 +48,25 @@ public Component getPageBodyComponent(String id) {
         return null;
     }
 
+    /** Tests if the data directory is embedded in the GeoServer web archive. */
+    public static boolean isEmbeddedDataDirectory(GeoServerApplication app) {
+        try {
+            String webRootPath = app.getServletContext().getRealPath("/");
+            if (webRootPath == null) {
+                return false; // this should never happen
+            }
+            File dataDir = app.getResourceLoader().getBaseDirectory().getAbsoluteFile();
```

Comment:
Thinking out loud here... what if the resource store is externalized in a database?
A data directory should be there anyways, at least the JDBC configuration files are in it, right? So there is a risk of losing configuration on upgrade if not externalized. Is it still a security risk though?

Reply:
I don't use the jdbcstore module but I did some investigation and even when using that module, a data directory still has to exist somewhere on the file system to tell geoserver how to connect to the jdbcstore database so both risks you mentioned would still exist.